lib/tasks/search.rake

```diff
@@ -1,170 +1,36 @@
-      #  Elasticsearch understands our document ids, so re-indexing the same
-      #  document won't create duplicates.
-      LOG.info "importing any documents that changed between #{start_time} and now"
-      cursor = Content.where(:_type.in => content_types, :updated_at.gte => start_time)
-      import_from_cursor(cursor, new_index, opts)
-    end
-  end
-
-  desc 'Copies contents of MongoDB into Elasticsearch if updated in the last N minutes.'
-  task :catchup, [:minutes, :batch_size, :sleep_time] => :environment do |t, args|
-    opts = batch_opts args
-    the_index = TaskHelpers::ElasticsearchHelper.get_index
-    alias_ = Tire::Alias.find the_index.name
-    # this check makes sure we are working with the index to which
-    # the desired model's alias presently points.
-    raise RuntimeError, "could not find live index" if alias_.nil?
+  desc 'Indexes content updated in the last N minutes.'
+  task :catchup, [:minutes] => :environment do |t, args|
     start_time = Time.now - (args[:minutes].to_i * 60)
-    cursor = Content.where(:_type.in => %w(Comment CommentThread), :updated_at.gte => start_time)
-    import_from_cursor(cursor, the_index, opts)
-  end
 
-  def batch_opts(args)
-    args = args.to_hash
-    {:batch_size => args[:batch_size].nil? ? 500 : args[:batch_size].to_i,
-     :sleep_time => args[:sleep_time].nil? ? 0 : args[:sleep_time].to_i}
+    [Comment, CommentThread].each do |model|
+      model.where(:updated_at.gte => start_time).import(index: Content::ES_INDEX_NAME)
+    end
   end
 
-  desc 'Removes any data from Elasticsearch that no longer exists in MongoDB.'
-  task :prune, [:batch_size, :sleep_time] => :environment do |t, args|
-    opts = batch_opts args
-    the_index = TaskHelpers::ElasticsearchHelper.get_index
-    puts "pruning #{the_index.name}"
-    alias_ = Tire::Alias.find the_index.name
-    raise RuntimeError, 'could not find live index' if alias_.nil?
-    scan_size = opts[:batch_size] / TaskHelpers::ElasticsearchHelper.get_index_shard_count(the_index.name)
-    cnt = 0
-    [CommentThread, Comment].each do |klass|
-      doc_type = klass.document_type
-      # this check makes sure we are working with the index to which
-      # the desired model's alias presently points.
-      search = Tire::Search::Scan.new the_index.name, {size: scan_size, type: doc_type}
-      search.each do |results|
-        es_ids = results.map(&:id)
-        mongo_ids = klass.where(:id.in => es_ids).map { |d| d.id.to_s }
-        to_delete = es_ids - mongo_ids
-        if to_delete.size > 0
-          cnt += to_delete.size
-          puts "deleting #{to_delete.size} orphaned #{doc_type} documents from elasticsearch"
-          the_index.bulk_delete (to_delete).map { |v| {"type" => doc_type, "id" => v} }
-        end
-        puts "#{the_index.name}/#{doc_type}: processed #{search.seen} of #{search.total}"
-        sleep opts[:sleep_time]
-      end
+  desc 'Reindex all data from the database'
```

Comment:
@feanil: The old rebuild (which is now reindex), used to do the following:
1. Create a new index with auto-generated name.
2. Special clean-up (i'll discuss later).
3. Load data into new index.
4. Move alias to above index.
5. Run slightly modified version of catch-up.

It sounds like you definitely want rebuild to do steps 1 and 3 above.  Do we also want to keep steps 4 and 5 in this single call as well, rather than having to make 3 separate calls (rebuild, move_alias, catchup) where inputs of future steps are dependent on outputs from earlier steps?  Keeping it together seems simpler for the user.  However, I'm not sure if @clintonb simplified it for a reason.

I had written an [earlier comment in the move_alias code](https://github.com/edx/cs_comments_service/pull/216/files#r93522259) regarding a special case where there is already an index named `content`, which is the name of the alias we wish to use, and where `move_alias` will fail.

It turns out that the original reindex code has special clean-up code for this situation (see line 57 of the original code in this file).  When it encounters this situation, it deletes the index named `content` and sets up the alias to the new index.

IMO - this clean-up code is still necessary, but I would make the following changes:
1. I wouldn't do this as step 2, but instead would do it as part of the other `move_alias`, after we have already loaded data into the newly created index, for obvious reasons.
2. I would add this clean-up functionality to `move_alias` with a force_delete parameter with default of false.  By default, move_alias would fail if it encounters this situation.  Within this rebuild, we would move the alias with force_delete=true.
3. It seems from the original code that the index could sometimes be re-created in between the time of the delete and the move.  I would additionally output the calls that should be made (move_alias with force_delete + catchup), in the event that this still fails.

Note that if we keep all of these tasks distinct, and don't try to make rebuild do the move and catch-up, it has the advantage of handling and noticing this clean-up as part of the full process.  I think I'd still advocate for the call to do everything, but just noting this advantage.